File: lib/tracker/locked-xp.ts
import type { CompletionType, Cadence, StatKind } from '../types';

export interface LockedXpInput {
  completion_type: CompletionType;
  target_value: number | null;
  cadence: Cadence;
}

/**
 * Auto-compute the locked base XP for a quest from its completion type, target,
 * and cadence. Pure: client-sent base_xp is always ignored in favor of this.
 *
 * - checkbox -> 25
 * - count    -> max(15, round(target_value * 0.5))
 * - timer    -> max(15, round(target_value * 0.85))
 * - weekly cadence multiplies the base by 1.5 (rounded)
 */
export function computeLockedXp({ completion_type, target_value, cadence }: LockedXpInput): number {
  let base: number;
  switch (completion_type) {
    case 'checkbox':
      base = 25;
      break;
    case 'count':
      base = Math.max(15, Math.round((target_value ?? 0) * 0.5));
      break;
    case 'timer':
      base = Math.max(15, Math.round((target_value ?? 0) * 0.85));
      break;
  }
  if (cadence === 'weekly') base = Math.round(base * 1.5);
  return base;
}

export function categoryXp(stat: StatKind): number {
  return stat === 'DIS' ? 20 : 10; // INT 10, STR 10, DIS 20
}

File: api/app/logic/locked_xp.py
"""Locked base-XP computation — port of lib/tracker/locked-xp.ts."""

from __future__ import annotations

from typing import Optional

from .types import CompletionType, StatKind


def compute_locked_xp(
    *, completion_type: CompletionType, target_value: Optional[int], cadence: str
) -> int:
    """Auto base XP from type/target/cadence. Client-sent base_xp is ignored.

    - checkbox -> 25
    - count    -> max(15, round(target * 0.5))
    - timer    -> max(15, round(target * 0.85))
    - weekly cadence multiplies the base by 1.5 (rounded)
    """
    target = target_value or 0
    if completion_type == "checkbox":
        base = 25
    elif completion_type == "count":
        base = max(15, _round_half_up(target * 0.5))
    elif completion_type == "timer":
        base = max(15, _round_half_up(target * 0.85))
    else:  # pragma: no cover - exhaustive over CompletionType
        raise ValueError(f"unknown completion_type {completion_type!r}")

    if cadence == "weekly":
        base = _round_half_up(base * 1.5)
    return base


def category_xp(stat: StatKind) -> int:
    return 20 if stat == "DIS" else 10  # INT 10, STR 10, DIS 20


def _round_half_up(x: float) -> int:
    """Match JavaScript Math.round (round half toward +Infinity)."""
    import math

    return math.floor(x + 0.5)
